### PR TITLE
Add network spawning for dynamic gltf-model objects. Make equippables work for network spawned objects.

### PR DESCRIPTION
**Pull request by HexaField**
## Summary

A summary of changes being made in this PR

## Checklist
- [ ] Pre-push checks pass `npm run check`
  - [ ] Linter passing via `npm run lint`
  - [ ] Unit & Integration tests passing via `npm run test:packages`
  - [ ] Docker build process passing via `npm run build-client`
- [ ] Changes have been manually QA'd 
- [ ] Changes reviewed by at least 2 approved reviewers

## References

References to pertaining issue(s)

## Reviewers

@HexaField @speigg @NateTheGreatt

### CODE REVIEW

File: packages/engine/src/interaction/functions/equippableFunctions.ts
Comment (NateTheGreatt) [inline packages/engine/src/interaction/functions/equippableFunctions.ts:16]: for my own understanding, is this `if` statement here to prevent the existing component data from being overwritten?

File: packages/engine/src/interaction/systems/EquippableSystem.ts
Comment (NateTheGreatt) [inline packages/engine/src/interaction/systems/EquippableSystem.ts:24]: what's the difference between `.matches` and `.matchesFromAny`? i always forget ><

File: packages/engine/src/interaction/systems/EquippableSystem.ts
Comment (NateTheGreatt) [inline packages/engine/src/interaction/systems/EquippableSystem.ts:101]:
```suggestion
```

File: packages/engine/src/interaction/systems/EquippableSystem.ts
Comment (NateTheGreatt) [inline packages/engine/src/interaction/systems/EquippableSystem.ts:77]:
```suggestion

```

File: packages/engine/src/interaction/systems/EquippableSystem.ts
Comment (NateTheGreatt) [inline packages/engine/src/interaction/systems/EquippableSystem.ts:64]:
```suggestion

```

File: packages/engine/src/interaction/systems/EquippableSystem.ts
Comment (NateTheGreatt) [inline packages/engine/src/interaction/systems/EquippableSystem.ts:60]:
```suggestion
```

File: packages/engine/src/scene/functions/SceneLoading.ts
Comment (NateTheGreatt) [inline packages/engine/src/scene/functions/SceneLoading.ts:108]:
```suggestion

```

File: packages/engine/src/scene/functions/SceneLoading.ts
Comment (NateTheGreatt) [inline packages/engine/src/scene/functions/SceneLoading.ts:118]:
```suggestion

```

File: packages/engine/src/networking/functions/incomingNetworkReceptor.ts
Comment (NateTheGreatt) [inline packages/engine/src/networking/functions/incomingNetworkReceptor.ts:64]:
```suggestion

```

File: packages/engine/src/networking/functions/incomingNetworkReceptor.ts
Comment (NateTheGreatt) [inline packages/engine/src/networking/functions/incomingNetworkReceptor.ts:71]:
```suggestion

```

File: packages/engine/src/networking/functions/incomingNetworkReceptor.ts
Comment (NateTheGreatt) [inline packages/engine/src/networking/functions/incomingNetworkReceptor.ts:44]:
```suggestion

```

File: packages/engine/src/networking/functions/incomingNetworkReceptor.ts
Comment (NateTheGreatt) [inline packages/engine/src/networking/functions/incomingNetworkReceptor.ts:42]:
```suggestion

```

File: packages/engine/src/interaction/systems/InteractiveSystem.ts
Comment (NateTheGreatt) [inline packages/engine/src/interaction/systems/InteractiveSystem.ts:69]: are these `hasComponent` checks necessary if it's removing the component anyways? if it's needed, perhaps we should add the `hadComponent` check to the `removeComponent` function instead

**Review by NateTheGreatt — CHANGES_REQUESTED** [review]
looks good, @hamzzam ! just a few questions/comments. the only changes i'd like to see before merging is to have those `hasComponent` calls moved into `removeComponent`. we might also want some of those `console.log` calls removed, too.

File: packages/engine/src/interaction/systems/EquippableSystem.ts
Comment (speigg) [reply packages/engine/src/interaction/systems/EquippableSystem.ts:24]: `.matches` will match an action only when it comes from a trusted source (e.g., the host, or itself)
`.matchesFromAny` will match an action no matter who sent it

Comment (speigg) [reply packages/engine/src/interaction/systems/EquippableSystem.ts:24]: @NateTheGreatt happy to rename these for clarity if you can think of better names

File: packages/engine/src/interaction/systems/InteractiveSystem.ts
Comment (speigg) [reply packages/engine/src/interaction/systems/InteractiveSystem.ts:69]: @NateTheGreatt Does bitecs allow `removeComponent` if the entity does not have the component?

Comment (hamzzam) [reply packages/engine/src/interaction/systems/InteractiveSystem.ts:69]: Yeah I was of the assumption that bitECS does not allow removeComponent if the component does not exist. But I just ran a test it does not throw any errors. So I will update the code and remove the extra hasComponent checks.

File: packages/engine/src/interaction/functions/equippableFunctions.ts
Comment (hamzzam) [reply packages/engine/src/interaction/functions/equippableFunctions.ts:16]: Yes.

File: packages/engine/src/interaction/systems/InteractiveSystem.ts
Comment (hamzzam) [reply packages/engine/src/interaction/systems/InteractiveSystem.ts]: Done

**Comment by hamzzam** [comment]
> looks good, @hamzzam ! just a few questions/comments. the only changes i'd like to see before merging is to have those `hasComponent` calls moved into `removeComponent`. we might also want some of those `console.log` calls removed, too.

Removed the console.logs and hasComponent checks. Let me know if anything else needs to be updated. Thanks

**Comment by HexaField** [comment]
Are you able to provide a guide to show the functionality for this? I tried using the collision cube to create an equippable and there is an issue with the new interactable UI and i cant seem to equip anything.

**Comment by hamzzam** [comment]
> Are you able to provide a guide to show the functionality for this? I tried using the collision cube to create an equippable and there is an issue with the new interactable UI and i cant seem to equip anything.

It was working fine before the latest dev merge. Let me try to debug

**Review by NateTheGreatt — APPROVED** [review]

**Review by NateTheGreatt — CHANGES_REQUESTED** [review]
QA steps would be great. i forgot to add that to the PR template

**Comment by hamzzam** [comment]
@HexaField @NateTheGreatt I tested it out, the equippables functionality is working fine. There are some errors in the interactables logic due to which the text and highlighting are not working. I tried adding some sanity checks but that didn't help. Probably the dev who worked on it will need to fix those.

To test the interactables you should do the following:

1. In the editor, add a model node and set its interaction type to Equippable. Save.
2. Then in the scene, take the avatar near the equippable model and press E. The object will attach to the avatar hand.

**Comment by NateTheGreatt** [comment]
thanks for the repro steps @hamzzam! I tested it out and there seems to be an issue with the item following the hand bone.

**Comment by hamzzam** [comment]
> thanks for the repro steps @hamzzam! I tested it out and there seems to be an issue with the item following the hand bone after pickup.

Hey @NateTheGreatt that's because the `getHandTransform` function is returning a fixed hand position currently. There is a TODO present to use the actual hand position when the animation rig is in. And I think that is in the system now, should I try fixing this? will need to look into the animation rig system. @HexaField

**Comment by HexaField** [comment]
> Hey @NateTheGreatt that's because the `getHandTransform` function is returning a fixed hand position currently. There is a TODO present to use the actual hand position when the animation rig is in. And I think that is in the system now, should I try fixing this? will need to look into the animation rig system. @HexaField

Yes I think we are now in a place where this should work.

**Comment by hamzzam** [comment]
> > Hey @NateTheGreatt that's because the `getHandTransform` function is returning a fixed hand position currently. There is a TODO present to use the actual hand position when the animation rig is in. And I think that is in the system now, should I try fixing this? will need to look into the animation rig system. @HexaField
> 
> Yes I think we are now in a place where this should work.

Ok. I will make a separate PR for that, we can get this PR approved if that is the only thing left.
Or let me know if you think this change must be included in this PR.

**Review by NateTheGreatt — APPROVED** [review]
> Ok. I will make a separate PR for that, we can get this PR approved if that is the only thing left.
Or let me know if you think this change must be included in this PR.

sounds good to me :+1:

**Pull request merged**
